pkg/cmd/checkendpoints/controller/controller.go

```diff
@@ -0,0 +1,255 @@
+	alpha1 "github.com/openshift/client-go/operatorcontrolplane/informers/externalversions/operatorcontrolplane/v1alpha1"
+	"github.com/openshift/client-go/operatorcontrolplane/listers/operatorcontrolplane/v1alpha1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/klog"
+
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/cmd/checkendpoints/operatorcontrolplane/podnetworkconnectivitycheck/v1alpha1helpers"
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/cmd/checkendpoints/trace"
+)
+
+// PodNetworkConnectivityCheckController continuously performs network connectivity
+// checks and reports the results.
+type PodNetworkConnectivityCheckController interface {
+	factory.Controller
+}
+
+// controller implements a PodNetworkConnectivityCheckController that discovers the list of endpoints to
+// check by looking for PodNetworkConnectivityChecks in a given namespace, for a specific pod. Updates to
+// the PodNetworkConnectivityCheck status are queued up and handled asynchronously such that disruptions
+// to the ability to update the PodNetworkConnectivityCheck status do not disrupt the ability to perform
+// the connectivity checks.
+type controller struct {
+	factory.Controller
+	podName      string
+	podNamespace string
+	checksGetter operatorcontrolplaneclientv1alpha1.PodNetworkConnectivityCheckInterface
+	checkLister  v1alpha1.PodNetworkConnectivityCheckNamespaceLister
+	recorder     events.Recorder
+	// each PodNetworkConnectivityCheck gets its own PodNetworkConnectivityCheckStatusUpdater
+	updaters map[string]PodNetworkConnectivityCheckStatusUpdater
+}
+
+// Returns a new PodNetworkConnectivityCheckController that performs network connectivity checks
+// as specified in the PodNetworkConnectivityChecks defined in the specified namespace, for the specified pod.
+func New(podName, podNamespace string, checksGetter operatorcontrolplaneclientv1alpha1.PodNetworkConnectivityChecksGetter, checkInformer alpha1.PodNetworkConnectivityCheckInformer, recorder events.Recorder) PodNetworkConnectivityCheckController {
+	c := &controller{
+		podName:      podName,
+		podNamespace: podNamespace,
+		checksGetter: checksGetter.PodNetworkConnectivityChecks(podNamespace),
+		checkLister:  checkInformer.Lister().PodNetworkConnectivityChecks(podNamespace),
+		recorder:     recorder,
+		updaters:     map[string]PodNetworkConnectivityCheckStatusUpdater{},
+	}
+	c.Controller = factory.New().
+		WithSync(c.Sync).
+		WithInformers(checkInformer.Informer()).
+		ResyncEvery(1*time.Second).
+		ToController("check-endpoints", recorder)
+	return c
+}
+
+// Sync ensures that the status updaters for each PodNetworkConnectivityCheck is started
+// and then performs each check.
```

Comment:
I expected this method to be

1. look to see if the set of networkchecks matched desired. if not add and start them.
2. remove ones that don't match.

In each network check, I expected to see
1. every second kick off an async call to `checkNetwork`
2. as the last call in `checkNetwork`, it calls `maybeUpdateStatus`. If there are 20 results or more since the last update, issue updateStatus

This code is mixing the two?  Can you describe what it's doing?

Reply:
### Overall:

1. There is goroutine per **PodNetworkConnectivityCheck** that maintains a queue of status updates and attempts to apply them once per second. This is the **PodNetworkConnectivityCheckStatusUpdater**.
2. The controller `Sync()`, triggered every second, that performs the network check and and queues status updates in the corresponding **PodNetworkConnectivityCheckStatusUpdater**.

### The `Sync` method:

1. Ensures a **PodNetworkConnectivityCheckStatusUpdater** is running for each **PodNetworkConnectivityCheck**. 
2. Stops any **PodNetworkConnectivityCheckStatusUpdater** that is no longer needed.
3. Calls `checkEndpoint()` for each check.
   > These should ideally be run async, but I'm having some trouble with that.

### The `checkEndpoint()` method:

1. performs the network check for a given **PodNetworkConnectivityCheck**.
2. queues the resultant updates in the corresponding  **PodNetworkConnectivityCheckStatusUpdater** instance.

### The **PodNetworkConnectivityCheckStatusUpdater**:

1. one instance is running per **PodNetworkConnectivityCheck**.
2. updates status (if needed) every second.